Refactor: register business hours block with block.json path

monsieur-z opened the pull request:
See https://github.com/Automattic/jetpack/issues/32602

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
Registers the _Business Hours_ block by specifying the path to its `block.json` file. This was made possible by https://github.com/Automattic/jetpack/pull/32697. The goal is to avoid the duplication of block metadata and use `block.json` as the source of truth.

In more details:
- Server side
  - Add a helper method to get the path to a block's metadata file (in the build folder)
  - Add tests
  - Register the block by specifying the path to `block.json`
- Client side
  - Remove `editorScript` from `block.json` file (automatically added during the build)
  - Move metadata from `index.js` to `block.json`
  - Delete `index.js`
  - Create a function that reads the icon from the metadata and generates a React component, in `shared-extension-utils`
  - Create a helper to register blocks by passing the metadata object
  - Register the block by passing the metadata
  - Write i18n-friendly CSS
  - Update block tests


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pedMtX-RS-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Test that the _Business Hours_ block behaves as in production.

<img width="682" alt="Screenshot 2023-09-13 at 3 50 35 PM" src="https://github.com/Automattic/jetpack/assets/1620183/6b61223a-e189-4b32-89b6-2f8a3c184ab9">


_Note: blocks metadata don't seem to be localized at the moment in production._

Commit:
[not verified] Enable block registration by specifying block.json path

Commit:
[not verified] Refactor: register business hours block with block.json path

monsieur-z commented on projects/plugins/jetpack/extensions/blocks/business-hours/block.json:
```diff
@@ -84,5 +84,64 @@
 			]
 		}
 	},
+	"example": {
```
From `index.js`.

monsieur-z commented on projects/plugins/jetpack/extensions/blocks/business-hours/business-hours.php:
```diff
@@ -22,23 +28,9 @@
+		get_block_path(),
 		array(
 			'render_callback' => __NAMESPACE__ . '\render',
-			'supports'        => array(
```
Data already present in `block.json`.

Commit:
[not verified] Make block CSS right-to-left compatible

monsieur-z commented on projects/plugins/jetpack/extensions/blocks/business-hours/style.scss:
```diff
@@ -8,7 +8,7 @@
 	dt {
 		font-weight: bold;
-		margin-right: 0.5em;
+		margin-inline-end: 0.5em;
```
Styles are still registered and enqueued manually in PHP. Eventually, they'll be determined from `block.json` and we won't be able to specify `view.css` or `view.rtl.css` at runtime. Let's write code that supports internationalization from the get-go.

jeherve replied:
Good catch!

Commit:
[not verified] Add test for registering block by its metadata file

Commit:
[not verified] Remove hardcoded editorScript value in business hours block.json

Commit:
[not verified] Move code to get block metadata path in Blocks

monsieur-z commented on projects/packages/blocks/src/class-blocks.php:
```diff
@@ -350,4 +350,20 @@ public static function is_standalone_block() {
+	 *
+	 * @return string
+	 */
+	public static function get_path_to_block_metadata( $block_src_dir ) {
```
I tried to add a test for that function, but I'm having trouble dealing with constants. Any insight would be appreciated.

monsieur-z replied:
Using `Constants::set_constant( 'JETPACK__PLUGIN_FILE', '/a/b/c/index.php' );` at the beginning of my test doesn't work. I end up with the following error: `Error: Undefined constant "Automattic\Jetpack\JETPACK__PLUGIN_FILE"`.

jeherve replied:
Right, this is a good indicator that we'd need to rework this, I think. `JETPACK__PLUGIN_FILE` will not work when folks want to use the method outside of Jetpack.
Maybe we could do something similar to this?
https://github.com/Automattic/jetpack/blob/3ff19e36551e19df74e26e87f9691846dfdd0be8/projects/packages/assets/src/class-assets.php#L136-L157

Commit:
[not verified] Move logic to render a block icon on the client

monsieur-z commented on projects/plugins/jetpack/extensions/blocks/business-hours/block.json:
```diff
@@ -84,5 +84,63 @@
 			]
 		}
 	},
-	"editorScript": "file:../editor.js"
```
Now generated automatically by the build step.

Commit:
[not verified] Remove deprecated file

monsieur-z commented on projects/plugins/jetpack/extensions/shared/get-block-icon-from-metadata.js:
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable jsdoc/no-undefined-types */
```
JSDoc doesn't like `React.Component` as a type.

Commit:
[not verified] Add changelogs

Commit:
[not verified] Fix failing test

Commit:
[not verified] Fix failing JS test

monsieur-z commented:
I did rebase the branch with `add/enable-block-registration-by-path`, but PHP tests are still failing here. Running `jetpack docker phpunit` locally from `projects/plugins/jetpack` doesn't show any failing test: I believe this may resolve itself once the base branch is merged.

jeherve commented on projects/plugins/jetpack/extensions/shared/get-block-icon-from-metadata.js:
What do you think about moving this to the shared-extension-utils package, so we can use it outside of the Jetpack plugin as well?

jeherve commented on projects/plugins/jetpack/extensions/shared/register-jetpack-block.js:
```diff
@@ -57,3 +58,24 @@ export default function registerJetpackBlock( name, settings, childBlocks = [],
+ * @param {boolean} prefix - See registerJetpackBlock.
+ * @returns {object|boolean} Either false if the block is not available, or the results of `registerBlockType`
+ */
+export function registerJetpackBlockFromMetadata( metadata, settings, childBlocks, prefix ) {
```
It's interesting, I would have expected a modified `registerJetpackBlock`, ready to handle different type of data passed to it, but I suppose a new function makes sense as well. 👍 It will probably be easier to understand that way!

jeherve commented on projects/packages/blocks/src/class-blocks.php:
```diff
@@ -352,4 +352,20 @@ public static function is_standalone_block() {
+	 * Returns the path to the directory (in the build folder) containing the block.json metadata
+	 * file of a block.
+	 *
+	 * @since 1.4.23
```
```suggestion
	 * @since $$next-version$$
```

Commit:
[not verified] Pass package build dir to get_path_to_block_metadata

Commit:
[not verified] Move block icon helpers to shared-extension-utils package

Commit:
[not verified] changelog

monsieur-z commented on projects/js-packages/shared-extension-utils/src/get-icon-color.js:
```diff
@@ -0,0 +1,25 @@
+import colorStudio from '@automattic/color-studio';
```
Duplication of `/projects/plugins/jetpack/extensions/shared/block-icons.js`. I'll remove that file once all blocks are using `getBlockIconComponent`.

Commit:
[not verified] Add missing import

Commit:
[not verified] Update package version

Commit:
[not verified] Add missing color-studio dependency to shared-extension-utils

Commit:
[not verified] Add missing jetpack-constants dep to blocks package

Commit:
[not verified] Remove call to realpath from get_block_metadata_from_file

Commit:
[not verified] Fix monorepo package version refs

Commit:
[not verified] Pass path to source block.json to jetpack_register_block

Commit:
[not verified] Fix wrong getIconColor import

monsieur-z commented on projects/packages/blocks/composer.json:
```diff
@@ -3,11 +3,13 @@
 		"automattic/wordbless": "dev-master",
 		"brain/monkey": "2.6.1",
-		"yoast/phpunit-polyfills": "1.1.0",
+		"yoast/phpunit-polyfills": "1.0.4",
```
Composer required a valid version.

jeherve replied:
This is odd. [1.1.0 is a valid version](https://github.com/Yoast/PHPUnit-Polyfills/releases/tag/1.1.0), and it's what we use all over the monorepo today.

monsieur-z commented on projects/plugins/jetpack/extensions/blocks/business-hours/business-hours.php:
```diff
@@ -12,33 +12,16 @@
 function register_block() {
 	Blocks::jetpack_register_block(
-		BLOCK_NAME,
+		__DIR__,
```
I had to replace `Blocks::get_path_to_block_metadata( __DIR__ )` (the path to the `block.json` file in the build folder) with `__DIR__` (the path to the `block.json` file in the source folder). The former implementation worked locally but made the test jobs in the CI pipeline fail: files in the build folder don't seem to exist at the time the tests are run. Please let me know if you have any insights about this.

The `block.json` files in the build folder include paths to scripts and stylesheets. Currently, block assets are registered and enqueued manually, so it doesn't matter which `block.json` we use for registration. But eventually, we'll need to use the one in the build folder.

Commit:
[not verified] Falls back to reading block.json in source directory

monsieur-z commented on projects/packages/blocks/src/class-blocks.php:
```diff
@@ -49,13 +50,16 @@ public static function jetpack_register_block( $slug, $args = array() ) {
 		if ( '/' === substr( $slug, 0, 1 ) ) {
-			$metadata = self::get_block_metadata_from_file( $slug );
-			$name     = self::get_block_name_from_metadata( $metadata );
+			// Test if build directory exists. Fall back to reading block.json from source directory.
```
The build folder for blocks (`_inc/block`) is not available when running the tests from the CI jobs. I'm still looking into why this is the case. In the meantime, we fall back to reading the `block.json` file from the block source directory so that tests pass. Hopefully, this is a temporary solution. Since block assets are currently registered and enqueued manually, reading from the `block.json` file in the build or source folder doesn't matter.

This logic is implemented here so that calling `jetpack_register_block` to register a block stays as straightforward as possible.

monsieur-z commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -478,9 +478,13 @@ public static function load_assets_as_required( $type, $script_dependencies = ar
+		// Retrieve the feature from block.json if a path is passed.
 		if ( '/' === substr( $type, 0, 1 ) ) {
-			$feature = Blocks::get_block_feature_from_metadata( Blocks::get_block_metadata_from_file( $type ) );
+			// Test if build directory exists. Fall back to reading block.json from source directory.
```
Same comment as for `jetpack_register_block`.

monsieur-z commented:
Since the last review, I:
- Moved the icon logic to the `shared-extension-utils` package.
- Updated the implementation of `get_path_to_block_metadata` by using `Jetpack_Constants`
- Added tests
- Passed the path to the block source directory to `jetpack_register_block`. As noted in the inline comment, that function takes care of reading the `block.json` file in the build folder, and falls back to the source folder.

jeherve commented on projects/js-packages/shared-extension-utils/package.json:
```diff
@@ -17,6 +17,7 @@
 		"test": "jest"
 	},
 	"dependencies": {
+		"@automattic/color-studio": "2.5.0",
```
It's a bit annoying that we have to bring that whole dependency just for a single color. Part of me would want to just hardcode the color like we do in other places like here:
https://github.com/Automattic/jetpack/blob/3a3c4adc1c0913bac224ddb832713bd6101fe1ab/projects/js-packages/components/components/jetpack-logo/index.tsx#L7

What do you think?

monsieur-z replied:
I'm okay with that.

Commit:
[not verified] Remove color studio dep in shared-extensions-util

Commit:
[not verified] Bump yoast/phpunit-polyfills version in blocks package

Commit:
[not verified] Update composer.lock in Jetpack plugin

Commit:
[not verified] Remove unhelpful changelogs

Commit:
[not verified] Add fallback logic to get_path_to_block_metadata

monsieur-z commented:
I updated the PR with the following:
- Removed the color studio dependency from the `shared-extension-utils` package and hardcoded the needed color instead.
- Forced version `1.1.0` of `yoast/phpunit-polyfills` after the fact, which didn't seem to cause any problem in the build pipeline.
- Moved the logic to handle a fallback `block.json` directory to `get_path_to_block_metadata` and updated the rest of the code accordingly.

monsieur-z commented on projects/packages/blocks/tests/php/test-blocks.php:
```diff
@@ -339,7 +340,7 @@ public function test_jetpack_register_block_with_existing_editor_style() {
 	 */
 	public function test_jetpack_register_block_from_metadata_file() {
-		$result = Blocks::jetpack_register_block( __DIR__ . '/fixtures/block.json' );
+		$result = Blocks::jetpack_register_block( __DIR__ . '/fixtures/test-block/block.json' );
```
Updated the file structure in `fixtures` for clarity and to facilitate `test_get_path_to_block_metadata`.

Commit:
Update package version

Commit:
Fix styling issues caused by span in getBlockIconComponent

fgiannar reviewed (commented):
Thanks for working on this, Kevin!

I tested this as follows: 

1. Added a business hours block on latest stable
2. Made a few changes and switched to the current branch
3. Confirmed the block is there and my changes persisted
4. Added a new business hours block and confirmed it behaves as expected
5. Confirmed no related errors in console

The only thing I found confusing (though doesn't look related to your changes) is that when editing the block the hours seem to be in my timezone but in view mode they appear in UTC
<img width="705" alt="Screenshot 2023-09-14 at 10 32 08" src="https://github.com/Automattic/jetpack/assets/1758399/84013723-6727-4293-b18a-45580be99975">

<img width="663" alt="Screenshot 2023-09-14 at 10 32 18" src="https://github.com/Automattic/jetpack/assets/1758399/9ec704cb-d006-48d0-b59d-ec3ba90ebfc7">

jeherve reviewed (approved):
Let's give this a try! 🚢
